classes/class-wc-connect-payment-methods-store.php

```diff
@@ -0,0 +1,128 @@
+				return;
+			}
+
+			$this->logger->log( 'Successfully loaded payment methods from server response.', __FUNCTION__ );
+			$this->update_last_fetch_timestamp();
+
+			// If we made it this far, it is safe to store the object
+			$this->update_payment_methods( $payment_methods );
+
+			// Potentially auto-update the selected method
+			$this->potentially_update_selected_payment_method_from_payment_methods( $payment_methods );
+		}
+
+		protected function potentially_update_selected_payment_method_from_payment_methods( $payment_methods ) {
+			$payment_method_count = count( $payment_methods );
+
+			// No payment methods at all? Clear anything we have stored
+			if ( 0 === $payment_method_count ) {
+				$this->service_settings_store->set_selected_payment_method_id( 0 );
+				return;
+			}
+
+			$payment_method_ids = array();
+			foreach ( (array) $payment_methods as $payment_method ) {
+				$payment_method_id = intval( $payment_method->payment_method_id );
+				if ( 0 !== $payment_method_id ) {
+					$payment_method_ids[] = $payment_method_id;
+				}
+			}
+
+			// Just one? Go ahead and use it
+			if ( 1 === $payment_method_count ) {
+				$this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
+				return;
+			}
+
+			// Is the stored method id not in the list? Select the first one
+			$selected_payment_method_id = $this->service_settings_store->get_selected_payment_method_id();
+			if ( ! in_array( $selected_payment_method_id, $payment_method_ids ) ) {
+				$this->service_settings_store->set_selected_payment_method_id( $payment_method_ids[ 0 ] );
+			}
+		}
+
+		public function get_payment_methods() {
+			return get_option( 'wc_connect_payment_methods', array() );
+		}
+
+		protected function update_payment_methods( $payment_methods ) {
+			update_option( 'wc_connect_payment_methods', $payment_methods );
+		}
+
+		public function get_last_fetch_timestamp() {
+			return get_option( 'wc_connect_payment_methods_last_update', null );
+		}
+
+		protected function update_last_fetch_timestamp() {
+			update_option( 'wc_connect_payment_methods_last_update', time() );
+		}
+
+		protected function get_payment_methods_from_response_body( $response_body ) {
+			if ( ! is_object( $response_body ) ) {
+				return new WP_Error( 'payment_method_response_body_type', 'Expected but did not receive object for response body.' );
+			}
+
+			if ( ! property_exists( $response_body, 'payment_methods' ) ) {
+				return new WP_Error( 'payment_method_response_body_missing_payment_methods', 'Expected but did not receive payment_methods in response body.' );
+			}
+
+			$payment_methods = $response_body->payment_methods;
+			if ( ! is_array( $payment_methods ) ) {
+				return new WP_Error( 'payment_methods_type', 'Expected but did not receive array for payment_methods.' );
+			}
+
+			foreach ( (array) $payment_methods as $payment_method ) {
+				$required_keys = array( 'payment_method_id', 'name', 'card_type', 'card_digits', 'expiry' );
```

Comment:
I would prefer to keep it. If we ever made a mistake on the server that caused one of these fields to be absent, I would not want to overwrite the option on the client with incomplete payment method data.